src/nxt_cgroup.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_malloc.h>
+#include <nxt_fs.h>
+#include <nxt_cgroup.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+
+
+static char *nxt_mk_cgpath(const char *dir);
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    int   ret, len;
+    FILE  *fp;
+    char  *cgpath;
+    char  cgprocs[PATH_MAX - 13];   /* - 13 for '/cgroup.procs' */
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.path == NULL)
+    {
+        return NXT_OK;
+    }
+
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgpath == NULL) {
+        return NXT_ERROR;
+    }
+    nxt_fs_mkdir_all((const u_char *) cgpath, 0777);
+
+    snprintf(cgprocs, sizeof(cgprocs), "%s/cgroup.procs", cgpath);
+    fp = fopen(cgprocs, "we");
+    if (fp == NULL) {
+        ret = NXT_ERROR;
+        goto out_free;
+    }
+
+    ret = NXT_OK;
+
+    setvbuf(fp, NULL, _IONBF, 0);
+    len = fprintf(fp, "%d\n", process->pid);
+    if (len < 0) {
+        ret = NXT_ERROR;
+    }
+
+    fclose(fp);
+
+out_free:
+
+    nxt_free(cgpath);
+
+    return ret;
+}
+
+
+void
+nxt_cgroup_cleanup(const nxt_process_t *process)
+{
+    char  *cgroot, *cgpath, *ptr;
+
+    cgroot = nxt_mk_cgpath("");
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgroot == NULL || cgpath == NULL) {
+        goto out_free;
+    }
+
+    while (*cgpath != '\0' && strcmp(cgroot, cgpath) != 0) {
+        rmdir(cgpath);
+        ptr = strrchr(cgpath, '/');
+        *ptr = '\0';
+    }
+
+out_free:
+
+    nxt_free(cgroot);
+    nxt_free(cgpath);
+}
+
+
+static char *
+nxt_mk_cgpath(const char *dir)
+{
+    char    *path, *buf, *ptr;
+    size_t  pathlen;
+
+    path = buf = ptr = NULL;
```

Comment:
I prefer not initializing variables if it's not strictly necessary.  Otherwise, we're silencing the "uninitialized variable is being used" warnings from the compiler, and instead silent bugs may happen.

Reply:
> I prefer not initializing variables if it's not strictly necessary. Otherwise, we're silencing the "uninitialized variable is being used" warnings from the compiler, and instead silent bugs may happen.

No, I agree. But in this case it's on purpose (you'll notice I didn't initialise the other variables in there, there is method to the madness!)

- If we don't do the if () we need ptr to be NULL.
- I set buf to NULL as
   1) I'm not sure if fscanf(3) will set it to NULL on failure
   2) If we don't do the if () or if we error out of it then we can still happily free(3) it
- path is set to NULL so we can just return that on error if we jumped to out_free_buf: (Yes, you could set path to NULL before doing the goto in the if ()... I chose to save a line of code...)

Reply:
- `ptr`: Yes, it needs to be nulled.
- `buf`: Yeah, this one too.
- `path`: Ahh, the goto forces us to initialize it too.  I tought this one wasn't necessary.

LGTM.